core-plugins/src/main/java/io/cdap/plugin/batch/aggregator/GroupByAggregator.java

```diff
@@ -131,9 +131,29 @@ public void validate(Schema inputSchema, List<String> groupByFields,
                                            functionInfo.getName(), functionInfo.getField()), null)
           .withConfigElement("aggregates", collectorFieldName);
       }
-    }
 
+      // TODO: CDAP-16401 - Push down validation to individual aggregate functions
+      if (GroupByConfig.Function.COUNTDISTINCT == functionInfo.getFunction()) {
+        if (!conf.containsMacro("aggregates")) {
+          validateCountDistinct(inputField, collector, collectorFieldName);
+        }
+      }
+    }
+  }
 
+  private void validateCountDistinct(Schema.Field inputField, FailureCollector collector, String validationFieldName) {
+    if (inputField != null) {
+      Schema.Type type = inputField.getSchema().isNullable() ?
+        inputField.getSchema().getNonNullable().getType() :
+        inputField.getSchema().getType();
+      if (type != Schema.Type.STRING && type != Schema.Type.INT && type != Schema.Type.LONG &&
+        type != Schema.Type.FLOAT && type != Schema.Type.DOUBLE && type != Schema.Type.BOOLEAN) {
```

Comment:
does it really make sense to allow float and double? It could potentially go out of memory easily, with 0.000001 different than 0.00000001, etc.

In Java when you check equality of floats and doubles most IDEs will flag it as a warning.

Reply:
though i guess the same could be set for CollectSet. We probably need a way to collect warnings and failures, with this falling under the warning category.

Reply:
Reported an improvement. https://issues.cask.co/browse/CDAP-16418. However, at this point, I don't think we need to absorb the risk of an OOM. I've removed support for FLOAT and DOUBLE right now. Can add it if someone brings up a use case explicitly.